feat: mention user in comments

sshanzel opened the pull request:
## Changes

Describe what this PR does
- created a common component for user's short info (image, name, username)
- once the user starts typing with at (@) character we will send a query and display the returned users
- when the user mentions someone, it must display differently by applying an additional style.

## Manual Testing

- On those affected packages:
- [x] Have you done sanity checks in the webapp?
- [ ] Have you done sanity checks in the extension?

- Did you test the modified components media queries?
- [ ] MobileL (420px)
- [ ] Tablet (656px)
- [x] Laptop (1020px)

DD-469 #done

sshanzel commented:
Should now be ready for review, just gotta work on those suddenly failing tests.

idoshamun commented on packages/shared/src/components/RecommendedMention.tsx:
```diff
@@ -0,0 +1,48 @@
+  }
+
+  return (
+    <ul
```
I'm not sure what should be the accessibility aspects of this component but you can have a look at Twitter and see if they added any attributes or keyboard behavior, etc. Please let me know your findings before you implement

sshanzel replied:
The noticeable thing on their structure looks to be utilizing a role attribute as `listbox` (select element) for the container and `option` role for each item. Then an aria attribute `aria-selected` for which one is currently selected (which now I think a must, to implement).

sshanzel replied:
For now, I will just apply the `aria-selected` attribute as it seems to be a must for this scenario. What do you think?

idoshamun replied:
I think the roles are also important so apply them as well, please.
They turn this component into a sort of input component, dropdown to be more accurate. So it's important to set these.

idoshamun commented on packages/shared/src/components/modals/NewCommentModal.tsx:
```diff
@@ -236,6 +258,8 @@ export default function NewCommentModal({
     ) {
       await sendComment(event);
     }
+
+    onMentionKeypress(event);
```
I think this is better

```suggestion
else {
    onMentionKeypress(event);
}  
```

idoshamun commented on packages/shared/src/components/modals/NewCommentModal.tsx:
```diff
@@ -316,13 +345,42 @@ export default function NewCommentModal({
         />
       </div>
+      {getShouldLoadTooltip() && (
+        <BaseTooltip
+          content={
+            <RecommendedMention
+              users={mentionQuery !== undefined && mentions}
+              selected={selected}
+              onClick={onMentionClick}
+            />
+          }
+          offset={[getOffsetPerChar(offsetX), getOffsetPerLine(offsetY) + -90]}
+          interactive
+          container={{
+            className: 'shadow',
+            paddingClassName: 'p-0',
+            roundedClassName: 'rounded-16',
+            bgClassName: 'bg-theme-bg-tertiary',
+          }}
+          reference={commentRef}
+          arrow={false}
+          placement="top-start"
+          visible={mentionQuery !== undefined}
+        />
```
This can go into its own component

idoshamun commented on packages/shared/src/components/modals/NewCommentModal.tsx:
```diff
@@ -316,13 +345,42 @@ export default function NewCommentModal({
       </div>
-      <footer className="flex justify-between items-center py-2 border-t border-theme-divider-tertiary">
+      <footer className="flex items-center py-3 px-3 border-theme-divider-tertiary">
+        <Button
+          className="mx-1 border border-theme-label-primary btn-tertiary"
+          buttonSize="small"
+          icon={<AtIcon />}
```
Why is this a button? What happens on click?

idoshamun commented on packages/shared/src/hooks/useProfileTooltip.ts:
```diff
@@ -0,0 +1,39 @@
+  data?: UserTooltipContentData;
+}
+
+export const useProfileTooltip = (userId: string): UseProfileTooltip => {
```
I'd like to suggest a different design instead of `fetchInfo`. You should provide another argument `enabled: boolean` to the hook and then you can provide it to the `enabled` prop of react query.
The current design seems somewhat like an anti pattern. What do you think?

sshanzel replied:
Was actually thinking about that, what made me do it this way is, that we won't have to manage an additional state outside for every consumer utilizing this hook.

idoshamun replied:
Ok then. Let's keep it this way :)

idoshamun commented on packages/shared/src/hooks/useUserMention.ts:
```diff
@@ -0,0 +1,205 @@
+      },
+    );
+  const { recommendedMentions: mentionUsers } = data;
+  const fetchUsers = useDebounce(refetch, 300);
```
Is this the same way we handle throttling on the search page? I think we should use the same mechanism

sshanzel replied:
It is currently using a timeout. Though their process flow is the same, should we update the search field implementations to utilize this hook?

idoshamun replied:
On a different ticket perhaps

sshanzel replied:
Alright, let me create one for it.

idoshamun commented on packages/shared/src/hooks/useUserMention.ts:
```diff
@@ -0,0 +1,205 @@
+    );
+  const { recommendedMentions: mentionUsers } = data;
+  const fetchUsers = useDebounce(refetch, 300);
+  const filteredUsers = mentionUsers?.filter((recommendation) =>
+    mentions.every((mention) => `@${recommendation.username}` !== mention),
+  );
```
What do you check here?

sshanzel replied:
Just so it won't have to, again, display the user that was already mentioned. Though after checking Twitter and GH, they allow users to be mentioned again in FE multiple times. In the backend though, it is already handled to be allowed. Let me abolish this.

idoshamun commented on packages/shared/src/hooks/useUserMention.ts:
```diff
@@ -0,0 +1,205 @@
+          query: mentionQuery,
+        }),
+      {
+        enabled: !!user,
```
You need to check here if `mentionQuery` is not `undefined` as well

idoshamun commented on packages/shared/src/hooks/useUserMention.ts:
```diff
@@ -0,0 +1,205 @@
+import { useQuery, useQueryClient } from 'react-query';
```
This file can use some more refactoring and better naming.
Some of the functions here are hard to follow and understand because there's quite a lot of complex logic here

idoshamun commented on packages/shared/src/components/Markdown.tsx:
```diff
@@ -1,12 +1,83 @@
+  const [username, setUsername] = useState('');
+  const [instance, setInstance] = useState<TippyInstance>();
+  const queryKey = ['profile', username];
+  const { data: fetchedProfile } = useQuery<PublicProfile>(
+    queryKey,
+    () => getProfile(username),
+    { enabled: !!username },
+  );
+  const { fetchInfo, data } = useProfileTooltip(fetchedProfile?.id);
```
Here we do double fetch. We run `getProfile` and inside `useProfileTooltip` we ask for the profile tooltip info.
Why can't we do only one? Maybe instead of filling the username in `data-mention-username` we should fill the user id. It will save a call. What do you think?

idoshamun reviewed (changes requested):
Wow man! This is brilliant engineering right there. Of course, I have some comments but still that's a massive work. Way to go 💪

rebelchris commented on packages/shared/src/components/RecommendedMention.tsx:
```diff
@@ -0,0 +1,48 @@
+  return (
+    <ul
+      className="flex overflow-hidden flex-col rounded-16 border border-theme-divider-secondary text-theme-label-primary"
+      style={{ width: '15rem' }}
```
Perhaps we can change this to be a JIT class?

rebelchris replied:
Ah yes probably, what mostly works is stopping the server, removing the .next file in your webapp and re-running.

sshanzel replied:
Worked! After re-run and deleting in-between. Thanks!

sshanzel replied:
@rebelchris looks like on our build it doesn't generate the arbitrary value. Have you had any situation where you were able to make it work in production?

Preview deployment:
![Screen Shot 2022-03-08 at 8 02 16 PM](https://user-images.githubusercontent.com/13744167/157234523-84cc7a0c-ca25-43e0-877c-3a23bbc343ee.png)

rebelchris replied:
I haven't seen it be an issue in production.
But don't let this block you, rather move back to style solution and add a separate ticket for this 😮‍💨

sshanzel replied:
Hmm. Let me clear my cache once I come back, could be just a cache issue on my end if it will be working on yours.

rebelchris commented on packages/shared/src/components/modals/NewCommentModal.tsx:
```diff
@@ -316,13 +346,42 @@ export default function NewCommentModal({
+              onClick={onMentionClick}
+            />
+          }
+          offset={[getOffsetPerChar(offsetX), getOffsetPerLine(offsetY) + -90]}
```
Does this not contradict when people are zoomed/using weird fonts? 
(not sure, might be worth checking)

sshanzel replied:
Looks like we're good to go! 

![Screen Shot 2022-03-07 at 10 37 46 PM](https://user-images.githubusercontent.com/13744167/157054913-e1b3cbe1-f8c1-42b4-b945-d4cc16ceba15.png)

rebelchris commented on packages/shared/src/hooks/useUserMention.ts:
```diff
@@ -0,0 +1,205 @@
+    // eslint-disable-next-line no-param-reassign
+    element.innerHTML = content;
+    setMentionQuery(undefined);
+    client.setQueryData(key, []);
```
Not a 100% sure, but you might also want to invalidate the query at this point.
Perhaps we can test it to see if the query keeps refreshing for instance?

sshanzel replied:
I agree a better way would be invalidating the query - You are right, it won't request again due to invalidation since the `mentionQuery` is already `undefined`. Though we won't be able to actually create a test for it as it resides inside a tooltip which we actually can't test since then. This is why I had to test the specific component on its own.

sshanzel replied:
Unfortunately, even though the `mentionQuery` is already undefined. I have to keep setting it as an empty array to avoid the additional request and flick on the UI when invalidating.

rebelchris reviewed (changes requested):
Wow Lee, really impressive work!
Looking forward to testing this out.

I added some small remarks from my side, let me know what you think

rebelchris reviewed (commented):
Also don't forget to add some test for anonymous users as the call needs auth.

sshanzel commented:
Should now be ready for another round of review! 🎉 🚢 

@idoshamun I have refactored the `useUserMention` and extracted some generic functions. With the filename though, I am not sure what to use for this one. Do you have one in mind?

@rebelchris we have a login barrier in our text comment modal popup, so anonymous users won't be able to type anything. I think the test for auth in this scenario shouldn't be necessary anymore. What do you think?

sshanzel commented:
Thank you both for the awesome review and kind words! 🙇  this should now be ready for review! 🎉 🚢 

@idoshamun I have now refactored the `useUserMention` and extracted the generic functions in a separate module. As for the filename, I'm not sure what to replace, do you have one in mind?

@rebelchris we have a login barrier for our comment modal popup, I think it should not be necessary anymore to test the anonymous user. What do you think?

sshanzel commented:
@rebelchris for testing different scenarios based on the sequence of letters. The result would be solely dependent on our API (integration testing). I think that scenario was covered in the API PR. Mocking the results I think won't provide much accuracy to whether the output was actually correct. What do you think?

idoshamun commented:
@sshanzel I think it's better to wait with the review until we will conclude the manual testing.
One point I did notice, we have some users with empty usernames because they didn't finish registration. We should exclude them. It's on the API side obviously.

rebelchris commented:
@sshanzel yep, don't worry about that test, we'll figure those out in the preview tests manually.

sshanzel commented:
Whew! This should now fix the edge cases for mentioning someone. Thanks a lot to Chris for listing out a lot of it as well. [Issues from this thread](https://dailydotdev.slack.com/archives/C01L0QXQJNB/p1646743316148079).

Should be ready for review 🚢

sshanzel commented:
This should now be ready for review 🙌 

cc @idoshamun @rebelchris @ayadotdev

rebelchris commented on packages/shared/src/components/Markdown.tsx:
```diff
@@ -1,12 +1,81 @@
+
+type TippyInstance = ReturnType<typeof tippy>[0];
+
+const getUserPermalink = (username: string) =>
+  `${process.env.NEXT_PUBLIC_WEBAPP_URL}${username}`;
```
Just thinking ahead, perhaps we can move this to a more generic file and export it?
This seems like something we'll need more often.

sshanzel replied:
I agree this should be in a different place.

rebelchris commented on packages/shared/src/components/modals/NewCommentModal.tsx:
```diff
@@ -265,7 +285,12 @@ export default function NewCommentModal({
+      {...{
+        contentRef: modalRef,
+        onRequestClose: confirmClose,
+        style: { content: { overflow: 'initial' } },
```
Do we need the overflow initial here? 
Could we not use `overflow-auto` here?

sshanzel replied:
In the past, we can't. I have now introduced a change that makes this unnecessary anymore 🚢

rebelchris commented on packages/shared/src/components/profile/ProfileTooltipContent.tsx:
```diff
@@ -42,11 +42,11 @@ export function ProfileTooltipContent({
         </p>
       )}
-      {tags?.length && (
+      {tags?.length ? (
```
Why do we need the hardcoded null return?

sshanzel replied:
It displays the string `0` when the length is 0 instead of conditionally rendering. If we turn it into a boolean, it might return the word `false`.

rebelchris reviewed (changes requested):
Getting so close, i'll do some UI testing now

idoshamun commented on packages/shared/src/components/Markdown.tsx:
```diff
@@ -1,12 +1,79 @@
+
+export default function Markdown({ content }: MarkdownProps): ReactElement {
+  const [userId, setUserId] = useState('');
+  const [instance, setInstance] = useState<TippyInstance>();
```
Why do we keep it in the state?
I see only where we assign a value to `instance` but not where we use it

sshanzel replied:
We will be using it in the very last part of generating the tooltip content. Below is line:

https://github.com/dailydotdev/apps/blob/fc87a2724404f9a9c0ef33e217fdca1c22f967f7/packages/shared/src/components/Markdown.tsx#L74

idoshamun commented on packages/shared/src/components/profile/UserShortInfo.tsx:
```diff
@@ -0,0 +1,55 @@
+        className,
+      )}
+    >
+      <ProfilePicture user={{ username, image }} size={imageSize} />
```
The following `{{ username, image }}` always creates a new object.
It's better to store it in `useMemo`

idoshamun commented on packages/shared/src/components/tooltips/RecommendedMentionTooltip.tsx:
```diff
@@ -0,0 +1,64 @@
+      reference={elementRef}
+      arrow={false}
+      placement="top-start"
+      visible={query !== undefined}
```
I'm not sure this is the right way to check for `undefined`. Please confirm it

sshanzel replied:
I think it would be much safer to check the type. Let me refactor it.

idoshamun commented on packages/shared/src/hooks/useUserMention.ts:
```diff
@@ -0,0 +1,231 @@
+      }
+
+      if (event.key === 'Backspace') {
+        await nextTick();
```
Why do we need this `nextTick` all over?

sshanzel replied:
I used to have this to utilize the `commentRef` as it is not updated the same time as when the user hits a key. The values are not matched. With that, I have reduced the usage and ensured things are still working as intended.

idoshamun replied:
Hopefully there's no race condition here

sshanzel replied:
Yes, I think there should be none, this is used now only at the initial phase.

idoshamun commented on packages/shared/src/hooks/useUserMention.ts:
```diff
@@ -0,0 +1,231 @@
+
+const ARROW_KEYS = ['ArrowUp', 'ArrowDown'];
+
+export function useUserMention({
```
The business logic here of how you handle the different chars like `Backspace` and other long code sections need more documentation I think. It will be hard to understand it in a few days.

sshanzel replied:
I have now added more information relating to this on the specs page in the product section at point number 5.

idoshamun commented on packages/shared/src/hooks/useUserMention.ts:
```diff
@@ -91,6 +92,10 @@ export function useUserMention({
   };
 
   const onMention = (username: string) => {
+    if (mentions.length === 0) {
```
Safer, I think :)

```suggestion
    if (!mentions?.length) {
```

idoshamun commented on packages/shared/src/components/tooltips/RecommendedMentionTooltip.tsx:
```diff
@@ -39,7 +39,7 @@ export function RecommendedMentionTooltip({
       content={
         <RecommendedMention
-          users={query !== undefined && mentions}
+          users={typeof query !== 'undefined' && mentions}
```
I think I saw more places where you checked for `undefined` like this `query !== undefined`.
Can you please validate?

idoshamun reviewed (changes requested):
Last review!

idoshamun reviewed (approved):
Yes!!!! Nailed it 💪

rebelchris reviewed (approved):
Nice work Lee 🛳

idoshamun reviewed (approved):
Much more simple now! I love it :)

rebelchris reviewed (approved):
Wow Lee, this is stunning so clean with the textareas!
From my side good to go 💖
